pkg/loghttp/query.go

```diff
@@ -35,6 +37,36 @@ const (
 type QueryResponse struct {
 	Status string            `json:"status"`
 	Data   QueryResponseData `json:"data"`
+	Links  []Link            `json:"links,omitempty"`
+}
+
+type Link struct {
+	Rel  string `json:"rel"`
+	Href string `json:"href"`
+}
+
+func NewSelfLink(r *RangeQuery) Link {
+	return Link{
+		Rel:  "self",
+		Href: fmt.Sprintf("/loki/api/v1/query_range?%s", r.Encode()),
+	}
+}
+
+func NewNextLink(request *RangeQuery, ts time.Time) Link {
+	q := request
+
+	// Based on the query direction we either set the start or end for the next query.
+	if q.Direction == logproto.FORWARD {
+		q.Start = ts.Add(1 * time.Nanosecond)
+	} else {
+		// The end timestamp is exclusive on a backward query.
+		q.End = ts
+	}
```

Comment:
This violates the comment in https://github.com/grafana/loki/blob/0afd113c966d8750c32a072fb94b861b294cbc7f/pkg/logcli/query/query.go#L133-L149. As I understand there could be two  entries on the same timestamp. If this case is at the end of a batch we might loose one entry unless we overlap the batches :see_no_evil: 

@owen-d and @slim-bean how often would this happen? I guess we have two choices:
1. Return overlapping batches and let callers de-duplicate.
2. Do not return overlapping batches but possibly not return all entries in a time range.

Reply:
It is happening !